### PR TITLE
utfjson support invalid encoding string

### PR DESCRIPTION
**Pull request by drmingdrmer**
add p3json, ported from python3-json. 不需要看
      add testsuite for python2 json to test p3json. 不需要看
      fix py2-json test suite to let it pass with py3-json module
      p3json add test cases to ensure unicode behavior
      utfjson: auto encode; encoding check, optional decoding when load().

这里是后3个点的diff,只看这个就可以了哈:
https://github.com/baishancloud/pykit/pull/45/files/52ae5ad2972b7c90ea97087d3816768986faec46..c335ba50a64060929592b976db89177ae1663d00

### CODE REVIEW

File: utfjson/utfjson.py [hunk 1]
```diff
@@ -1,16 +1,118 @@
...
+    if isinstance(o, dict):
+        rst = {}
+        for k, v in o.items():
+            rst[encode_unicode(k, encoding)] = encode_unicode(v, encoding)
+
+    elif isinstance(o, list):
+        rst = []
+        for v in o:
+            rst.append(encode_unicode(v, encoding))
+    else:
```
Comment (shuoqingding) [inline utfjson/utfjson.py]: 对于`set`、`Iterator`、`tuple`之类的类型，在 utfjson 里应该不会用到。但作为公开的接口，还是应该处理一下（直接不支持报个错）？ `ensure_unicode` 和 `decode` 也有这个问题。

File: utfjson/utfjson.py [hunk 2]
```diff
@@ -1,16 +1,118 @@
...
+def dump(obj, encoding='utf-8'):
+
+    # - Using non-unicode with ensure_ascii=True results in unexpected output:
+    #       utfjson.dump('我',  ensure_ascii=True)  --> '"\\u00e6\\u0088\\u0091"'
+    #
+    #   We do not allow this to happen.
+    #
+    #
+    # - Using nonunicode with ensure_ascii=False results in unexpected output:
+    #       utfjson.dump(u'我', ensure_ascii=False) --> u'"\u6211"'
```
Comment (shuoqingding) [inline utfjson/utfjson.py]: `p3json.dump`?

File: utfjson/utfjson.py [hunk 3]
```diff
@@ -1,16 +1,118 @@
 #!/usr/bin/env python2
 # coding: utf-8
 
-import json
+from pykit import p3json
```
Comment (shuoqingding) [inline utfjson/utfjson.py]: 建议加一点注释说一下为啥要用 p3json。

**Comment by shuoqingding** [comment]
只通过测试的话，不知道会不会有别的问题。灰度发布的时候得多观察一下。

File: p3json/__init__.py [hunk 4]
```diff
@@ -338,7 +338,8 @@ def loads(s, encoding=None, cls=None, object_hook=None, parse_float=None,
     The ``encoding`` argument is ignored and deprecated.
 
     """
-    if isinstance(s, str):
+    # XXX changes from python2 json: basestring is base class of str and unicode
```
Comment (drmingdrmer) [inline p3json/__init__.py]: XXX就是一个标记(这里小心的意思)..跟TODO差不多,更没意义1点..

好我统一撸1下

File: p3json/test/test_dump.py [hunk 5]
```diff
@@ -15,8 +15,12 @@ def test_encode_truefalse(self):
         self.assertEqual(self.dumps(
                  {True: False, False: True}, sort_keys=True),
                  '{"false": true, "true": false}')
+        # XXX changes from python2 json: no long int support
+        # self.assertEqual(self.dumps(
+        #         {2: 3.0, 4.0: 5L, False: 1, 6L: True}, sort_keys=True),
+        #         '{"false": 1, "2": 3.0, "4.0": 5, "6": true}')
```
Comment (drmingdrmer) [inline p3json/test/test_dump.py]: 本来是想保留原来的代码的内容. 能看出来和原始的python3的实现的差别.

File: utfjson/utfjson.py (hunk 1 above)
Comment (drmingdrmer) [inline utfjson/utfjson.py]: 貌似处理下tuple支持就可以了. 其他的,`set`, `Iterator` 可以直接忽略吧? 下层真正dumps的时候会报错

**Comment by drmingdrmer** [comment]
@shuoqingding 是的. 应该要观察久1点. 有1个变化, load进来的对象里的字符串默认都不是unicode了, 是str. 这个和之前是不1样的. 可能有些地方要手动指定转unicode.

File: p3json/test/test_dump.py (hunk 5 above)
Comment (shuoqingding) [inline p3json/test/test_dump.py]: 我的意思是，注释掉的部分应该可以通过吧？`no long int support` 不是被你修复了么？

**Comment by shuoqingding** [comment]
要是有个 Feature Toggle 的话，能放心不少：http://featureflags.io/python-feature-flags/

**Comment by shuoqingding** [comment]
可以考虑让来实习俩月的哥们做一个这个东西。还有 web 版的 errlog 也可以考虑。

File: p3json/test/test_dump.py (hunk 5 above)
Comment (drmingdrmer) [inline p3json/test/test_dump.py]: 擦..好像是过了..

**Comment by drmingdrmer** [comment]
猴啊..可以..

**Comment by drmingdrmer** [comment]
image
